pkg/provision/azure.go

```diff
@@ -0,0 +1,375 @@
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2017-05-10/resources"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/google/uuid"
+	"github.com/sethvargo/go-password/password"
+	"log"
+)
+
+type AzureProvisioner struct {
+	subscriptionId    string
+	resourceGroupName string
+	deploymentName    string
+	authorizer        autorest.Authorizer
+	ctx               context.Context
+}
+
+func NewAzureProvisioner(subscriptionId string) (*AzureProvisioner, error) {
+	authorizer, err := auth.NewAuthorizerFromFile(azure.PublicCloud.ResourceManagerEndpoint)
+	ctx := context.Background()
+	return &AzureProvisioner{
+		subscriptionId: subscriptionId,
+		authorizer:     authorizer,
+		ctx:            ctx,
+	}, err
+}
+
+// Provision provisions a new Azure instance as an exit node
+func (p *AzureProvisioner) Provision(host BasicHost) (*ProvisionedHost, error) {
+
+	log.Printf("Provisioning host with Azure\n")
+
+	p.resourceGroupName = "inlets-" + host.Name
+	p.deploymentName = "inlets-deploy-" + uuid.New().String()
+
+	log.Printf("Creating resource group %s", p.resourceGroupName)
+	group, err := createGroup(p, host)
+	if err != nil {
+		return nil, err
+	}
+	log.Printf("Resource group created %s", *group.Name)
+
+	log.Printf("Creating deployment %s", p.deploymentName)
+	err = createDeployment(p, host)
+	if err != nil {
+		return nil, err
+	}
+	return &ProvisionedHost{
+		IP:     "Creating",
+		ID:     p.resourceGroupName,
+		Status: ActiveStatus,
+	}, nil
+}
+
+// Status checks the status of the provisioning Azure exit node
+func (p *AzureProvisioner) Status(id string) (*ProvisionedHost, error) {
+	deploymentsClient := resources.NewDeploymentsClient(p.subscriptionId)
+	deploymentsClient.Authorizer = p.authorizer
+	deployment, err := deploymentsClient.Get(p.ctx, p.resourceGroupName, p.deploymentName)
+	if err != nil {
+		return nil, err
+	}
+	var deploymentStatus string
+	if *deployment.Properties.ProvisioningState == "Succeeded" {
+		deploymentStatus = ActiveStatus
+	} else {
+		deploymentStatus = *deployment.Properties.ProvisioningState
+		if deploymentStatus == "Running" {
+			deploymentStatus = "deploying"
+		}
+	}
+	IP := "Creating"
+	if deploymentStatus == ActiveStatus {
+		IP = deployment.Properties.Outputs.(map[string]interface{})["publicIP"].(map[string]interface{})["value"].(string)
+	}
+	return &ProvisionedHost{
+		IP:     IP,
+		ID:     id,
+		Status: deploymentStatus,
+	}, nil
+}
+
+// Delete deletes the Azure exit node
+func (p *AzureProvisioner) Delete(request HostDeleteRequest) error {
+	groupsClient := resources.NewGroupsClient(p.subscriptionId)
+	groupsClient.Authorizer = p.authorizer
+	groupDeleteFuture, err := groupsClient.Delete(p.ctx, request.ID)
+	if err != nil {
+		return err
+	}
+	log.Printf("Waiting for deletion completion (~10 mins)")
```

Comment:
Please do not wait, this is counter to the way the provisioning package must work, it must return as soon as possible.